### PR TITLE
#9645 Implement FTP extensions for IPv6

### LINKED ISSUE #9645: twisted.protocols.ftp.FTP should support IPv6 extensions
|[cjwatson's avatar](https://github.com/cjwatson)| @cjwatson reported|
|-|-|
|Trac ID|trac#9645|
|Type|enhancement|
|Created|2019-05-31 13:52:32Z|

RFC 2428 defines FTP extensions needed to make it work over IPv6, which Twisted doesn't yet support.  This [blocks ppa.launchpad.net being available over IPv6](https://bugs.launchpad.net/txpkgupload/+bug/1807662) (well, we'll probably monkey-patch support rather than actually waiting for a Twisted release, but we still need to do most of the same work).

This is related to [#9085](https://github.com/twisted/twisted/issues/9085), but that's the client side whereas this is the server side.

<details><summary>Searchable metadata</summary>

```
trac-id__9645 9645
type__enhancement enhancement
reporter__cjwatson cjwatson
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__ftp ftp
keywords__None None
time__1559310752044062 1559310752044062
changetime__1570997590451628 1570997590451628
version__None None
owner__cjwatson cjwatson

```
</details>

### PR DESCRIPTION
**Pull request by cjwatson**
Fixes #9645

These are defined in RFC 2428.

The facilities in the RFC for requesting the use of a specific network
protocol don't fit very well into Twisted's view of the world, but I've
at least arranged for the command arguments to be validated
appropriately for the common cases of IPv4 and IPv6.

Tests for IPv6 support are only partial, because FTPClient doesn't yet
support this (see https://twistedmatrix.com/trac/ticket/9085).

### CODE REVIEW

**Comment by cjwatson** [comment]
There are some remaining lint errors, but they're essentially all due to the justification of "=" characters in constant assignments near the top of the file, and mostly pre-existing except where I made my changes fit the surrounding code.  I think it would make more sense to waive these rather than clutter this PR with them, but if reviewers require it then I can do the reformatting.

File: src/twisted/protocols/ftp.py
Comment (adiroiban) [inline src/twisted/protocols/ftp.py]: Do we need these members as public members?

File: src/twisted/protocols/ftp.py
Comment (adiroiban) [inline src/twisted/protocols/ftp.py]: I think that is ok to return a failure here...

What I did, is create a standard reponse code like 

```
RESPONSE[EPSV_ALL_ACTIVE] = '503 EPSV ALL was enabled.'
```

and then just do

```
        if self._epsvAllActive:
            return (EPSV_ALL_ACTIVE,)
```

File: src/twisted/protocols/ftp.py
Comment (adiroiban) [inline src/twisted/protocols/ftp.py]: This is ok.

Maybe the error can be:  `PASV available only for IPv4` ... or something more verbose for the client :)

Adding support for mapped IPV4 is nice, but I don't know if it is worth adding all this extra handling here.

What I have found for many of my deployments is that with FTP servers behind the load balances, people will want to hardcode the Load Balancer IP in the PASV response... so that old clients will continue to work.

File: src/twisted/protocols/ftp.py
Comment (glyph) [inline src/twisted/protocols/ftp.py]: Thanks so much for proactively adding docs for this!

File: src/twisted/protocols/ftp.py
Comment (glyph) [inline src/twisted/protocols/ftp.py]: Appreciate that you made this a tuple so that we wouldn't get cruft accumulating in it by accident :)

File: src/twisted/protocols/ftp.py
Comment (glyph) [inline src/twisted/protocols/ftp.py]: Can you add a `@param` explaining its usage, since this is technically modifying a public API?

Also, `''` is an ambiguous default value; is this supposed to be bytes or text?

File: src/twisted/protocols/ftp.py
Comment (glyph) [inline src/twisted/protocols/ftp.py:1010]: Minor stylistic note: the best exception-handling form here would be to be *even more* constrained here; i.e.

```
coercedHost = _coercedUnicode(host)
try:
    address = IPv6Address(coercedHost)
except ...:
    pass
```

Mainly just a note for future reference.

File: src/twisted/test/test_ftp.py
Comment (glyph) [inline src/twisted/test/test_ftp.py]: `test_conch` and `test_udp` both already define functions with *exactly* this name, and almost exactly this same implementation. 3 total repetitions of this code is too many :).  Please refactor it into a common (private, for now) location, so we're not making 3 different useless sockets during the test suite.

File: src/twisted/test/test_ftp.py
Comment (glyph) [inline src/twisted/test/test_ftp.py]: ahem, `_hasIPv4MappedAddresses` :)

File: src/twisted/test/test_ftp.py
Comment (glyph) [inline src/twisted/test/test_ftp.py:13]: For what it's worth, these tests are in the old, bad style.  The new, good style would be to consistently use `twisted.internet.testing.MemoryReactor` and not require real, actual IPv6 local connectivity in order to cover this code.  I realize that this isn't really a new problem in this branch, so it doesn't need to be addressed by this branch, but I feel like I should really bring it up to get the word out.

File: src/twisted/test/test_ftp.py
Comment (glyph) [inline src/twisted/test/test_ftp.py:1573]: Can you add a test docstring explaining what this covers more precisely?

File: src/twisted/test/test_ftp.py
Comment (glyph) [inline src/twisted/test/test_ftp.py]: "definitely" is a strong word here :).

File: src/twisted/test/test_ftp.py
Comment (glyph) [inline src/twisted/test/test_ftp.py]: that fires when...?

File: src/twisted/test/test_ftp.py
Comment (glyph) [inline src/twisted/test/test_ftp.py]: Can you describe this in terms of what's being verified rather than what the setup does?  ( https://jml.io/pages/test-docstrings.html is the canonical guideline. )

File: src/twisted/python/_setup.py
Comment (glyph) [inline src/twisted/python/_setup.py]: Thanks for correctly updating the package metadata for this requirement.  It strikes me as slightly unfortunate that this is our first `ipaddress` import anywhere, since we consistently use `twisted.internet.address` everywhere else.  Since I'm sending this back to you for feedback, maybe *consider* adding the IPv4 mapped test to our own class?  I wouldn't say it should be a hard requirement, since we probably transitively depend on this six different ways anyway now, and it is in the stdlib, but it does stick out just a little.

If you consider it and think it's the best choice, then by all means leave it :)

**Review by glyph — CHANGES_REQUESTED** [review]
The changes requested are pretty minor; for the most part this is a great change.

Regarding the lint thing: when I'm looking for reviews to do, I tend to hit the ones with green checkmarks first, so I'd probably go ahead and make the style inconsistent to appease the mechanical checks.  It's not a dealbreaker, but it'll make it easier and quicker to complete the review.  Maybe we should do a lint-ratchet reduction PR to reformat the rest of the surrounding stuff but let's not bundle those together.

**Review by adiroiban — COMMENTED** [review]
I didn't had time to finalize the review...but  I am sending my current comments.

**Comment by cjwatson** [comment]
I filed #1190 to clean up lint, and will return to this once that's merged.

File: src/twisted/protocols/ftp.py
Comment (cjwatson) [reply src/twisted/protocols/ftp.py:1010]: Effectively done by merging current Twisted trunk, which dropped Python 2 compatibility and thus the `_coercedUnicode` call just goes away.

File: src/twisted/protocols/ftp.py
Comment (cjwatson) [reply src/twisted/protocols/ftp.py]: I've added the `@param`.  The ambiguous default was an issue when you said this, but is no longer ambiguous now that Twisted only supports Python 3.

File: src/twisted/python/_setup.py
Comment (cjwatson) [reply src/twisted/python/_setup.py]: I agree this is novel, and I've considered the change you suggest.  I think if `twisted.internet.address` had any interrogation properties or methods rather than being pure data containers (with a few dunder methods), my answer might be different.  As it is, it seems really difficult to justify the effort and extra API surface of adding this to `twisted.internet.address` when there's already an adequate facility in the stdlib, especially now that we require Python 3 and so we no longer need an extra dependency for it.

So I prefer to leave this as it is, although I've dropped the setup change since `python_version < '3.0'` is now always true for Twisted.

File: src/twisted/protocols/ftp.py
Comment (cjwatson) [reply src/twisted/protocols/ftp.py]: I suppose not.  I've made them private now.

File: src/twisted/protocols/ftp.py
Comment (cjwatson) [reply src/twisted/protocols/ftp.py]: I suppose I could have done, but I was following the existing pattern set by `ftp_PASV` and `ftp_PORT`, and it didn't really seem worth lifting this up to the top level when each message is only used in one place.

File: src/twisted/protocols/ftp.py
Comment (cjwatson) [reply src/twisted/protocols/ftp.py]: I've made the error message clearer.

Not supporting IPv4-mapped addresses caused a [regression](https://bugs.launchpad.net/txpkgupload/+bug/1836059) in our environment, which is why I added this.  It's a relatively small amount of extra code compared to `EPSV` as a whole.

**Comment by cjwatson** [comment]
@glyph @adiroiban Sorry for taking nearly two years to get round to dealing with your review comments!  I think I've dealt with pretty much everything now, apart from the `MemoryReactor` comment (which I agree with but I don't have the energy to refactor all the FTP tests at this point, and also as @glyph says it's not a new problem in this branch), and the `narrativedocs` test failures which are fixed in #1667, so this should be ready for another round of review now.

**Comment by glyph** [comment]
No need to apologize — if anything, managing to finally come back after 2 years rather than just dropping it forever is praiseworthy :).

**Review by adiroiban — APPROVED** [review]
Looks good. Thanks.

I think that this can be merged as it's already a big improvement over trunk.

----------

 for my project I am using a highly modified version of twisted FTP protocol with support for py2 and Implicit/Explict FTPS, x509 based authentication, and machine listing and a few other extensions . client and server side.

if such functionality is needed in twisted and someone has time to review the changes, let me know and I can look at sending them for review

**Comment by adiroiban** [comment]
I am enabled auto-merge. It looks like  the merge is blocked as @glyph  has rejected the changes.

Once the changes are approved by @glyph the PR should be automatically merged.

Thanks

**Comment by adiroiban** [comment]
@glyph can  we merge this? Do you still think that this needs changes? Thanks!

File: src/twisted/python/_setup.py
Comment (glyph) [reply src/twisted/python/_setup.py]: > So I prefer to leave this as it is, although I've dropped the setup change since `python_version < '3.0'` is now always true for Twisted.

always False for Twisted, I hope :)

**Review by glyph — APPROVED** [review]

**Comment by glyph** [comment]
In the future, you can always feel free to dismiss my reviews if you've reviewed the points that I raised and they've been addressed.

**Pull request merged**
